### ISSUE TITLE
Add an interface to configure checkpoints

### ISSUE THREAD

**Issue by ferki** [melezhik/sparrow#2]
It would probably be nice to have a sparrow command that helps with configuring checkpoints, e.g. something like:

``` bash
$ sparrow setconfig df-check --threshold=80
```

Of course, the above is just an example, not an exact syntax requested :)

The idea can also be extended to have a command to get the actual configuration values too.

**Comment by ferki** [comment]
Maybe `sparrow check set $project $check $option $value` can already be used or extended to do  something like this? As in `sparrow check set system df-check threshold 80`?

**Comment by melezhik** [comment]
HI Ferenc! Initially sparrow `check set` command was designed to bind a given checkpoint to sparrow plugin , plus setting optionally a hostname ( for those plugins where hostname parameter  makes a sense ) , probably it's a good idea to evolve 'check set' command to add any other parameters to check point / plugin pair , I need to think about it. One thing that I am a bit concerned in this way that then we would have 2 different commands for defining sparrow checkpoint settings - `check set` and `check ini, load ini` which might be confusing ...

**Comment by melezhik** [comment]
From other hand, it will not be a problem, as long as we use these 2 commands in different context:
- `check set` - to set check point  ini setting automatically ( via rex for example )
- `check set ini` - to set checkpoint ini settings via user's sessions - manually by user input ... 

So using `sparrow check set` to set checkpoint / plugin parameters at runtime does make a sense ...

**Comment by melezhik** [comment]
Hi Ferenc , here is my further thoughts on  setting sparrow checkpoint parameters ...

Do we really need to set checkpoint parameters  so they **update checkpoint ini file**? 

Let's imagine situation when we have different values for `foo` parameter in both cmdb and by calling  `sparrow check set --foo=bar` command,  obviously once  `foo` being set by  `sparrow check set` command , at the next run of `Misc:Sparrow:setup` task it will be restored to one stored in cmdb.

IMHO we need only one single source of check point configuration ( cmdb or sparrow.yml file ) when we talk about rex/sparrow integration to avoid confusions.

Indeed instead of this I would rather talk about setting parameter for sparrow check point **during run time**, something like that 

`rex Misc:Sparrow:Check --check system.disk --threshold=60`

This task just will override default threshold value on per session basis and once check is done the default value for threshold still remain in ini file.

PS Initial idea behind this issue was to provide sparrow interface to populate checkpoint ini files by input data provided by rex ( yml files ) , sorry If lead away the discussion :smile: ,  should we put this task to dedicated github issue or keep using this one?

**Comment by ferki** [comment]
Well, my point here was to avoid any tools outside sparrow scope (including, but not limited to rex) have to know about the internal configuration structure of sparrow (which is currently INI), and how to manipulate that (generate from template, search and replace keys/values, etc.).

Instead, I proposa that any non-sparrow utility should be enabled to just "don't care", and ask sparrow itself to configure its own plugins on-demand. After all, sparrow should always know how the configuration should be done.

For example, should sparrow choose to replace the INI format with YAML (or database, or anything else), then all exeternal tools (including, but not limited to rex), would be broken until they relase a new version.

Where it is related to rex integration, is this: if there would be such interface for sparrow, I would have avoided "being smart" and generate INI files with rex for sparrow to consume. That creates a stronger coupling between them, and it might be wise to avoid that in the long run.

So it's not urgent or something, it's just nice to have to make it simpler to interface with sparrow :)

> at the next run of Misc:Sparrow:setup task it will be restored to one stored in cmdb.

Well, that's the _purpose_ of a configuration management tool: to enforce settings that are described in CMDB. If someone alters the configuration manually and does not record the change in the CMDB, then I don't see much point for them to use configuration management software in the first place.

> we need only one single source of check point configuration ( cmdb or sparrow.yml file ) when we talk about rex/sparrow integration to avoid confusions.

Exactly, if one decides to configure sparrow with rex, they should not apply configuration changes without rex (i.e. manually on the remote hosts).

> setting parameter for sparrow check point during run time

That also sounds like a good idea :+1: But I think it's separate from this original topic here.

**Comment by melezhik** [comment]
Hi Ferenc! Ok, I see what you mean. Let me think a little bit of the best way to implement such an interface ... From one hand I want to keep sparrow configuration simple ( as just editing ini files  ) and from other hand need to provide better integration with external tools  where I need to take into account formats like json or yml. Providing a intermediate layer could be one of the way, I just don't want to have it too complicated, so I need to come to some compromise ...  

> setting parameter for sparrow check point during run time

will be done in #3

**Comment by melezhik** [comment]
Hi Ferenc!

After some mediation I came up with the idea of 2 ways configuration system for sparrow plugins.
One - for manual sparrow invocation and other for automatic sparrow invocation, like we have with rex.

Let me explain a bit.

| configuration type | sparrow command | internal format | data source | when applied? | customer |
| --- | --- | --- | --- | --- | --- |
| manual | `sparrow check ini / sparrow load_ini` | .Ini style | checkpoint related ini file | before checkpoint run | a human being |
| external | `sparrow check run --yaml /path/to/conifg/file.yaml` | YAML | yml file gets populated as rex CMDB file | during checkoint run | rex |

Now if talk about plugin configuration introspection:

| configuration type | introspection command | customer |
| --- | --- | --- |
| manual | `sparrow check show` | a human being |
| external | `dump of CMDB yaml file` | rex |

As both yaml and .Ini format a _kinda_ compatible I don't think it would be an issue to have 
a dual configuration ( INI/YAML ) for one sparrow plugin. Internally both INI/YAML will be converted into hopefully the same Perl hash. Consider df-check plugin:
- .INi style : 

```
[disk]
threshold = 80
```
- YAML : 

```
---
disk:
    threshold: 80
```

Both variants will be `config()->{disk}->{threshold}` object ...

Let me know what do you think?  An updated documentation of how this could be done on sparrow side is here - https://github.com/melezhik/sparrow#configuring-checkpoints 

Thanks ...

**Comment by melezhik** [comment]
As for the rex CMDB files it will requires using a _sub_tree_ of yml data to pass into `sparrow check run --yaml sparrow.yml` , for example: 

default.yml ===>  sparrow.sytstem.checkname.settings ==> sparrow.yml

```
$ cat cmd/default.yml

sparrow:
    system:
        - checkname: disk
          plugin: df-check
          settings:
              threshold: 44
        - checkname: nginx
          plugin: nginx-check
          settings:
              check_master_age: 0
              master_age: 60
        - checkname: nginx-404-errors
          plugin: logdog
          settings:
              file: /var/log/nginx/access.log
              time_pattern: \[(\d+\/\S+\/\d+):(\S+)
              time_format: %d/%b/%Y %T
              history: 5 minutes
              timezone: UTC
              filter: HTTP\/\S+?"\s+404\s
              key_field: (\S+)
```

**Comment by ferki** [comment]
I find the idea of different config files of different formats applied to different scopes confusing. How about coming back to this topic, when the first problems arise around the current implementation?

I assumed if sparrow knows how to read its own ini config files, it would be very easy to teach it to also write them (instead of teaching every external tool, including rex, how to write sparrow config files). Apparently my assumption was wrong :)

It can be kept open if you consider implementing config file writing support in sparrow later, but you can also close it if it's not going to happen in the foreseeable future.

**Comment by melezhik** [comment]
>  How about coming back to this topic, when the first problems arise around the current implementation?

Indeed I have encountered issue for so called [private](https://github.com/melezhik/sparrow#private-plugins) sparrow plugins. As rex relies on internal path for sparrow plugin default ini file, those paths are differ for private and public plugins:
- public - `File::Spec->join( '~', 'sparrow', 'plugins', 'public', $check->{plugin},`
- private - `File::Spec->join( '~', 'sparrow', 'plugins', 'private', $check->{plugin},`

This is why I suggested the idea of sparrow would consume yaml configurations provided by rex ( instead of current approach with relies on inter sparrow structure like knowing ini files paths and making regular expression substitution , so on )

Other points for this:
- yml if native for rex, so it is not going to be much efforts for integration with sparrow
- yml is of  widely usage , why not support it on sparrow side?
- both Ini style and YAML formats are _kinda_ compatible as long as the are converted into perl hashes which is internal representation for configurations in sparrow

> I find the idea of different config files of different formats applied to different scopes confusing

I would agree,  it is a bit confusing , but as long as we rely only on rex CMDB as data source and don't mix them with ini files approach , this is not an issue ...

>  How about coming back to this topic, when the first problems arise around the current implementation?

Basically  yes , by see my comment concerning private plugins , the current rex/sparrow integration works for public plugins though, probably it is enough to let people to take it for a spin and we can change this latter if need ...

**Comment by ferki** [comment]
> Basically yes , by see my comment concerning private plugins

That's exactly why I keep saying that external tools should not be concerned about how sparrow handles its plugins and their configuration internally. That's clearly sparrow's territory, others only could emulate it at best.

What I proposed here originally is to provide a subcommand of sparrow to make it possible to manipulate plugin settings without the need to know how it is implemented inside sparrow.

Like if I execute `sparrow config set system df-check threshold 60` then sparrow would write the proper ini-file contents to the proper location, and apply them.

I feel like we massively overengineered this simple idea here.

**Comment by melezhik** [comment]
closing this due to https://github.com/RexOps/rex-recipes/pull/29
